File: jvm/src/test/kotlin/com/jetbrains/signatureverifier/tests/authenticode/Authenticode.kt
package com.jetbrains.signatureverifier.tests.authenticode

import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.x509.DigestInfo

open class SpcIndirectDataContent(private val data: SpcAttributeOptional, private val messageDigest: DigestInfo) :
  ASN1Encodable {
  override fun toASN1Primitive(): ASN1Primitive {
    return BERSequence(getAsn1EncodableVectorInstance(data, messageDigest))
  }
}

open class SpcAttributeOptional(private val type: ASN1ObjectIdentifier, private val value: ASN1Encodable?) :
  ASN1Encodable {
  override fun toASN1Primitive(): ASN1Primitive {
    val v = getAsn1EncodableVectorInstance(type)
    if (value != null)
      v.add(value)
    return BERSequence(v)
  }
}

open class SpcPeImageData : ASN1Encodable {
  private val flags = DERBitString(ByteArray(0))
  private val file = SpcLink()

  override fun toASN1Primitive(): ASN1Primitive {
    return BERSequence(getAsn1EncodableVectorInstance(flags, DERTaggedObject(0, file)))
  }
}

open class SpcLink : ASN1Encodable, ASN1Choice {
  private val file = SpcString("")
  override fun toASN1Primitive(): ASN1Primitive {
    return DERTaggedObject(false, 2, file)
  }
}

open class SpcString(str: String) : ASN1Encodable, ASN1Choice {
  private val unicode = DERBMPString(str)

  override fun toASN1Primitive(): ASN1Primitive {
    return DERTaggedObject(false, 0, unicode)
  }
}

private fun getAsn1EncodableVectorInstance(vararg asn1Encodable: ASN1Encodable): ASN1EncodableVector {
  return ASN1EncodableVector().also { it.addAll(asn1Encodable) }
}

File: net/JetBrains.SignatureVerifier/tests/Authenticode.cs
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X509;

namespace JetBrains.SignatureVerifier.Tests
{
  public class SpcIndirectDataContent : Asn1Encodable
  {
    private readonly SpcAttributeOptional myData;
    private readonly DigestInfo myMessageDigest;

    public SpcIndirectDataContent(SpcAttributeOptional data, DigestInfo messageDigest)
    {
      myData = data;
      myMessageDigest = messageDigest;
    }

    public override Asn1Object ToAsn1Object()
    {
      return new BerSequence(new Asn1EncodableVector(myData, myMessageDigest));
    }
  }

  public class SpcAttributeOptional : Asn1Encodable
  {
    private readonly DerObjectIdentifier type;
    private readonly Asn1Encodable value;

    public SpcAttributeOptional(DerObjectIdentifier type, Asn1Encodable value)
    {
      this.type = type;
      this.value = value;
    }

    public override Asn1Object ToAsn1Object()
    {
      var v = new Asn1EncodableVector(type);
      if (value != null) v.Add(value);
      return new BerSequence(v);
    }
  }

  public class SpcPeImageData : Asn1Encodable
  {
    private DerBitString flags = new(new byte[0]);
    private SpcLink file = new();

    public override Asn1Object ToAsn1Object()
    {
      return new BerSequence(new Asn1EncodableVector(flags, new DerTaggedObject(0, file)));
    }
  }

  public class SpcLink : Asn1Encodable, IAsn1Choice
  {
    private SpcString file = new("");

    public override Asn1Object ToAsn1Object()
    {
      return new DerTaggedObject(false, 2, file);
    }
  }

  public class SpcString : Asn1Encodable, IAsn1Choice
  {
    private DerBmpString unicode;

    public SpcString(string str)
    {
      unicode = new DerBmpString(str);
    }

    public override Asn1Object ToAsn1Object()
    {
      return new DerTaggedObject(false, 0, unicode);
    }
  }
}